truffleruby/src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -507,32 +516,155 @@ private Object getMatchData() {
+
+        private NullStrategy() {
+            super(null, null, null);
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            return null;
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocalBlock(RubyContext context, DynamicObject block) {
+            return null;
+        }
+
+        @Override
+        public boolean matches(Frame callerFrame) {
+            return callerFrame == null;
+        }
+    }
+
+    public static class SimpleSetLastMatchStrategy extends SetLastMatchStrategy {
+        public SimpleSetLastMatchStrategy(FrameDescriptor fd, FrameSlot fs, Object defaultValue) {
+            super(fd, fs, defaultValue);
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            return getThreadLocalObjectFromFrame(context, callerFrame, fs, defaultValue, true);
+        }
+    }
+
+    public static class ComplexSetLastMatchStrategy extends SetLastMatchStrategy {
+        private final int depth;
+        public ComplexSetLastMatchStrategy(FrameDescriptor fd, FrameSlot fs, Object defaultValue, int depth) {
+            super(fd, fs, defaultValue);
+            this.depth = depth;
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            final MaterializedFrame frame = RubyArguments.getDeclarationFrame(callerFrame, depth);
+            return getThreadLocalObjectFromFrame(context, frame, fs, defaultValue, true);
+        }
+    }
+
     @Primitive(name = "regexp_set_last_match", needsSelf = false)
     @ImportStatic(RegexpNodes.class)
     public static abstract class RegexpSetLastMatchPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
+        @Child ReadCallerFrameNode readCallerFrame = new ReadCallerFrameNode(CallerFrameAccess.READ_WRITE);
+
         public static RegexpSetLastMatchPrimitiveNode create() {
             return RegexpSetLastMatchPrimitiveNodeFactory.create(null);
         }
 
-        public abstract DynamicObject executeSetLastMatch(Object matchData);
+        public abstract DynamicObject executeSetLastMatch(VirtualFrame frame, Object matchData);
 
-        @TruffleBoundary
-        @Specialization(guards = "isSuitableMatchDataType(getContext(), matchData)")
-        public DynamicObject setLastMatchData(DynamicObject matchData) {
-            Frame frame = getContext().getCallStack().getCallerFrameIgnoringSend().getFrame(FrameInstance.FrameAccess.READ_WRITE);
-            ThreadLocalObject lastMatch = getMatchDataThreadLocalSearchingDeclarations(
-                    getContext(), frame, true);
+        @Specialization(guards = { "isSuitableMatchDataType(getContext(), matchData)",
+                "strategy.matches(readCallerFrame.execute(frame))" }, limit = "20")
```

Comment:
No magic numbers please! Add an option (see `tools/options.yaml` and `jt build options`).